internal/config/config.go

```diff
@@ -57,35 +67,93 @@ type Config struct {
 	SidecarRequestsMemory string            `default:"40Mi" desc:"Lower bound of the NSM sidecar requests memory limits (in k8s resource management units)" split_words:"true"`
 	SidecarRequestsCPU    string            `default:"100m" desc:"Lower bound of the NSM sidecar requests CPU limits (in k8s resource management units)" split_words:"true"`
 	envs                  []corev1.EnvVar
+	secretsClient         coreV1Types.SecretInterface
 	caBundle              []byte
 	cert                  tls.Certificate
+	mode                  Mode
 	once                  sync.Once
 }
 
+// Mode type
+type Mode uint32
+
+// These are the different mode of webhook setup.
+const (
+	// SelfregisterMode allows you to use an automatically generated webhook configuration and certificate
+	SelfregisterMode Mode = iota
+	// SpireMode requires using spire configuration to obtain certificate and manually applying webhook configuration
+	SpireMode
+	// SecretMode requires to use k8s tls secret from the same Config.Namespace with the provided certificates
+	SecretMode
+)
+
+// These are the expecting fields name in k8s certificate secret
+const (
+	certFieldName = "tls.crt"
+	keyFieldName  = "tls.key"
+)
+
 // GetOrResolveEnvs converts on the first call passed Config.Envs into []corev1.EnvVar or returns parsed values.
-func (c *Config) GetOrResolveEnvs() []corev1.EnvVar {
-	c.once.Do(c.initialize)
+func (c *Config) GetOrResolveEnvs(ctx context.Context) []corev1.EnvVar {
+	c.once.Do(func() { c.initialize(ctx) })
 	return c.envs
 }
 
-// GetOrResolveCertificate tries to create certificate from Config.CertFilePath, Config.KeyFilePath or creates self signed in memory certificate.
-func (c *Config) GetOrResolveCertificate() tls.Certificate {
-	c.once.Do(c.initialize)
-	return c.cert
+// GetOrResolveMode tries to parse Config.WebhookMode and return parsed values.
+func (c *Config) GetOrResolveMode(ctx context.Context) Mode {
+	c.once.Do(func() { c.initialize(ctx) })
+	return c.mode
 }
```

Comment:
Im not sure that we should do spire cert reteriving operation in the config package. Is it possible to do it in the main package?

Reply:
Moved cert retrieving operation to the main